execution_engine/src/shared/transform.rs

```diff
@@ -168,23 +170,42 @@ impl Transform {
     /// Applies the transformation on a specified stored value instance.
     ///
     /// This method produces a new [`StoredValue`] instance based on the [`Transform`] variant.
+    ///
+    /// This method will panic if self is a [`Transform::Delete`] variant.
     pub fn apply(self, stored_value: StoredValue) -> Result<StoredValue, Error> {
+        match self.apply_optional(stored_value) {
+            Ok(Some(new_value)) => Ok(new_value),
+            Ok(None) => {
+                // Delete transform can't be handled here as it implies a stored value is present.
+                // Delete transforms should be handled before applying effects on stored values to
+                // avoid an unnecessary global state read.
+                unreachable!("Delete operation can't be applied");
+            }
+            Err(error) => Err(error),
+        }
+    }
+    /// Applies the transformation on a specified stored value instance.
+    ///
+    /// This method produces a new [`StoredValue`] instance based on the [`Transform`] variant. If a
+    /// given transform is a [`Transform::Delete`] then `None` is returned as the [`StoredValue`] is
+    /// consumed but no new value is produced.
+    fn apply_optional(self, stored_value: StoredValue) -> Result<Option<StoredValue>, Error> {
```

Comment:
I don't think we should have a separate method only to handle `Transform::Delete`s without panicking.

I think it would be much clearer if `apply` were just updated to return the optional stored value and we updated the few callsites to handle this.  Probably in many(all?) cases we could just panic at the callsite if `Ok(None)` is returned, but I feel this is more future-proof/less footgunny than putting the panicking code inside `apply`.

Reply:
Done as discussed: `Transform::apply` now returns `Option<StoredValue>` where Some indicate newly produced value, and None means no new value was produced and the value should be deleted from the global state.